core/src/com/google/inject/internal/FailableCache.java

```diff
@@ -16,65 +16,103 @@
 
 package com.google.inject.internal;
 
-import com.google.common.cache.CacheBuilder;
-import com.google.common.cache.CacheLoader;
-import com.google.common.cache.LoadingCache;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Maps;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
 
 /**
  * Lazily creates (and caches) values for keys. If creating the value fails (with errors), an
  * exception is thrown on retrieval.
  *
+ * This defends against re-entrant loads, and removals of elements while they are being loaded.
+ *
  * @author jessewilson@google.com (Jesse Wilson)
  */
 public abstract class FailableCache<K, V> {
 
-  private final LoadingCache<K, Object> delegate =
-      CacheBuilder.newBuilder()
-          .build(
-              new CacheLoader<K, Object>() {
-                @Override
-                public Object load(K key) {
-                  Errors errors = new Errors();
-                  V result = null;
-                  try {
-                    result = FailableCache.this.create(key, errors);
-                  } catch (ErrorsException e) {
-                    errors.merge(e.getErrors());
-                  }
-                  return errors.hasErrors() ? errors : result;
-                }
-              });
+  private final Map<K, FutureTask<V>> delegate = new LinkedHashMap<>();
```

Comment:
"curious why this needs to be a linked map instead of just a plain map.  if it doesn't need to be linked, we could get rid of explicit synchronization using ConcurrentHashMap instead AFAICT.

(only question of switching to concurrent would be atomicity of calling create(..) during the computeIfAbsent->newFutureTask.  unsure if atomic, unsure if needs to be.)"